File: codes/backend/app/schemas/state.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


SessionStatus = Literal["not_started", "active", "paused", "takeover", "ended"]
StateEventType = Literal["state_reset", "instructor_cue"]


class Vitals(BaseModel):
    heart_rate: int
    spo2: int
    respiratory_rate: int
    blood_pressure: str


class Symptoms(BaseModel):
    breathing_effort: str
    chest_tightness: str
    cough: str
    dizziness: str


class Emotion(BaseModel):
    anxiety: str
    fatigue: str


class VoiceBehavior(BaseModel):
    speech_pattern: str
    tone: str


class Interventions(BaseModel):
    oxygen_applied: bool
    bronchodilator_given: bool
    provider_notified: bool
    patient_repositioned: bool


class SafetyControls(BaseModel):
    ai_paused: bool = False
    instructor_takeover: bool = False


class PatientState(BaseModel):
    scenario_id: str
    status: SessionStatus = "active"
    stage: str
    vitals: Vitals
    symptoms: Symptoms
    emotion: Emotion
    voice_behavior: VoiceBehavior
    interventions: Interventions
    safety: SafetyControls = Field(default_factory=SafetyControls)
    last_updated_at: datetime


class StateEvent(BaseModel):
    event_id: str
    event_type: StateEventType
    timestamp: datetime
    state_after: PatientState
    cue_id: str | None = None
    label: str | None = None


class AutoPatientMessage(BaseModel):
    message_id: str
    speaker: Literal["patient"] = "patient"
    text: str
    trigger: Literal["instructor_cue"] = "instructor_cue"
    cue_id: str
    cue_label: str | None = None


class PatientStateResponse(BaseModel):
    state: PatientState
    auto_patient_message: AutoPatientMessage | None = None


class StateEventsResponse(BaseModel):
    events: list[StateEvent]

File: codes/frontend/src/api/state.ts
import { API_BASE_URL } from "./client";

export type SessionStatus =
  | "not_started"
  | "active"
  | "paused"
  | "takeover"
  | "ended";

export type StateEventType =
  | "state_reset"
  | "instructor_cue"
  | "pause"
  | "resume"
  | "takeover_started"
  | "takeover_ended";

export type Vitals = {
  heart_rate: number;
  spo2: number;
  respiratory_rate: number;
  blood_pressure: string;
};

export type Symptoms = {
  breathing_effort: string;
  chest_tightness: string;
  cough: string;
  dizziness: string;
};

export type Emotion = {
  anxiety: string;
  fatigue: string;
};

export type VoiceBehavior = {
  speech_pattern: string;
  tone: string;
};

export type Interventions = {
  oxygen_applied: boolean;
  bronchodilator_given: boolean;
  provider_notified: boolean;
  patient_repositioned: boolean;
};

export type SafetyControls = {
  ai_paused: boolean;
  instructor_takeover: boolean;
};

export type PatientState = {
  scenario_id: string;
  status: SessionStatus;
  stage: string;
  vitals: Vitals;
  symptoms: Symptoms;
  emotion: Emotion;
  voice_behavior: VoiceBehavior;
  interventions: Interventions;
  safety: SafetyControls;
  last_updated_at: string;
};

export type StateEvent = {
  event_id: string;
  event_type: StateEventType;
  timestamp: string;
  state_after: PatientState;
  cue_id: string | null;
  label: string | null;
};

export type AutoPatientMessage = {
  message_id: string;
  speaker: "patient";
  text: string;
  trigger: "instructor_cue";
  cue_id: string;
  cue_label: string | null;
};

export type PatientStateResponse = {
  state: PatientState;
  auto_patient_message: AutoPatientMessage | null;
};

export type StateEventsResponse = {
  events: StateEvent[];
};

export async function getPatientState(): Promise<PatientStateResponse> {
  const response = await fetch(`${API_BASE_URL}/state`);

  if (!response.ok) {
    throw new Error(`State request failed with status ${response.status}`);
  }

  return response.json();
}

export async function resetPatientState(): Promise<PatientStateResponse> {
  const response = await fetch(`${API_BASE_URL}/state/reset`, {
    method: "POST",
  });

  if (!response.ok) {
    throw new Error(`State reset failed with status ${response.status}`);
  }

  return response.json();
}

export async function applyInstructorCue(
  cueId: string,
): Promise<PatientStateResponse> {
  const response = await fetch(`${API_BASE_URL}/state/cues/${cueId}`, {
    method: "POST",
  });

  if (!response.ok) {
    throw new Error(`Cue request failed with status ${response.status}`);
  }

  return response.json();
}

export async function getStateEvents(): Promise<StateEventsResponse> {
  const response = await fetch(`${API_BASE_URL}/state/events`);

  if (!response.ok) {
    throw new Error(`State events request failed with status ${response.status}`);
  }

  return response.json();
}

export async function pauseAiPatient(): Promise<PatientStateResponse> {
  return postStateSafetyControl("/state/safety/pause", "AI pause");
}

export async function resumeAiPatient(): Promise<PatientStateResponse> {
  return postStateSafetyControl("/state/safety/resume", "AI resume");
}

export async function startInstructorTakeover(): Promise<PatientStateResponse> {
  return postStateSafetyControl(
    "/state/safety/takeover/start",
    "Instructor takeover start",
  );
}

export async function endInstructorTakeover(): Promise<PatientStateResponse> {
  return postStateSafetyControl(
    "/state/safety/takeover/end",
    "Instructor takeover end",
  );
}

async function postStateSafetyControl(
  path: string,
  label: string,
): Promise<PatientStateResponse> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    method: "POST",
  });

  if (!response.ok) {
    throw new Error(`${label} request failed with status ${response.status}`);
  }

  return response.json();
}
